### PR TITLE
Further reduce quad count of generated item models 

### PR DESCRIPTION
**Pull request by bs2609**
Followup PR to #4332.

This is implemented conditionally, as unlike the previous PR, there are some tradeoffs in rendering quality.

In particular, overlaps can be seen in the body of translucent models, as shown here:

* Current:
quads_normal

* With changes enabled:
quads_reduced

This also doesn't work well with the changes made by 62d0f012361a05e9a6a5bf71f58fe4d5b1b1df59, so this also switches that code to use the previous values when needed.

`UnpackedBakedQuad` numbers:
* Forge 2498+: 15,476 instances, 19.45MB heap usage
* With PR changes: 10,224 instances, 12.85MB heap usage

Heap usage of `BakedItemModel` falls from 20.30MB to 13.66MB, a further 6.64MB saving (32.7%)

Quad count changes for vanilla models can be found [here](https://gist.github.com/bs2609/437c0ce0ede04cf90b2eefb11362bc80).

The change is to only generate a single quad spanning all visible pixels in a single row or column's sides. The current method generates quads for each contiguous span, with gaps inbetween. (The old method generated a quad for each visible pixel in a side.)

The altered approach here is basically the approach taken by the vanilla item model generator, and produces the minimum number of quads needed to render the whole model. It does however lead to quads intersecting and overlapping each other, which is where the potential for rendering artifacts comes from.

Vanilla also appears to handle translucency differently (see #4401). Forge's items do look somewhat better as a result, but it makes the interior overlaps visible here.

This PR also includes a test mod, which just contains a few test items generated from various texure patterns.

**Update**:
These changes are now applied except when processing models for textures with translucency.

### CODE REVIEW

**Comment by bs2609** [comment]
It would also be possible to always use the existing approach for models made from textures with translucency, as they have the most visible problems.

**Comment by bs2609** [comment]
Another note:

The config option is currently set to require a restart. This isn't strictly needed, but at present there's no "requires resource reload" option, which is what would be required here.

**Comment by LexManos** [comment]
That looks horrible, no...

**Comment by mezz** [comment]
With side effects like that, these changes are entering the territory of performance mods like Optifine and not really 'safe' enough for Forge. 
Is there any way to detect a scenario where it will cause issues, and switch to the other logic?

**Comment by bs2609** [comment]
That is a test case specifically designed to show those negative effects. Here's what the changes look like otherwise:

* Current:


* With changes enabled:


Also see my [comment](https://github.com/MinecraftForge/MinecraftForge/pull/4448#issuecomment-334745546), this could be disabled for translucent textures, as they'll have the visible effects.

**Comment by mezz** [comment]
If you implement this as an optimization for non-translucent textures (i.e. most of them) then I think this would be well worth considering.

File: src/main/java/net/minecraftforge/client/model/ItemLayerModel.java [hunk 1]
```diff
@@ -356,21 +398,28 @@ private int getIndex(int u, int v)
...
 
-    private static boolean isTransparent(int[] pixels, int uMax, int vMax, int u, int v)
+    private static int getAlpha(int[] pixels, int uMax, int vMax, int u, int v)
     {
-        return (pixels[u + (vMax - 1 - v) * uMax] >> 24 & 0xFF) == 0;
+        return pixels[u + (vMax - 1 - v) * uMax] >> 24 & 0xFF;
     }
 
-    private static BakedQuad buildSideQuad(VertexFormat format, Optional<TRSRTransformation> transform, EnumFacing side, int tint, TextureAtlasSprite sprite, int u, int v, int size)
+    private static boolean isTransparent(int alpha)
```
Comment (mezz) [inline src/main/java/net/minecraftforge/client/model/ItemLayerModel.java]: This method should be inlined, it's kinda overkill.

File: src/main/java/net/minecraftforge/common/ForgeModContainer.java [hunk 2]
```diff
@@ -113,6 +113,7 @@
     public static boolean disableVersionCheck = false;
     public static boolean forgeLightPipelineEnabled = true;
     public static boolean replaceVanillaBucketModel = true;
+    public static boolean minimizeItemModelQuads = false;
```
Comment (mezz) [inline src/main/java/net/minecraftforge/common/ForgeModContainer.java]: I don't think the config is necessary now that there are no artifacts.

**Review by mezz — CHANGES_REQUESTED** [review]

File: src/main/java/net/minecraftforge/common/ForgeModContainer.java (hunk 2 above)
Comment (bs2609) [reply src/main/java/net/minecraftforge/common/ForgeModContainer.java]: There's still the matter of the changes in 62d0f012361a05e9a6a5bf71f58fe4d5b1b1df59 no longer being applied.

Comment (mezz) [reply src/main/java/net/minecraftforge/common/ForgeModContainer.java]: What is that commit fixing exactly, and why does this PR reintroduce the problem?

File: src/main/java/net/minecraftforge/client/model/ItemLayerModel.java (hunk 1 above)
Comment (bs2609) [reply src/main/java/net/minecraftforge/client/model/ItemLayerModel.java]: I was planning on proposing a change to `isTransparent`, so I kept it easily changeable here.

**Comment by bs2609** [comment]
Regarding mezz's question about quad offsets, this PR doesn't change the problem, it's just that the solution applied there doesn't really work with the changes made here. See:

* PR changes + post-(62d0f012361a05e9a6a5bf71f58fe4d5b1b1df59) logic:


* PR changes + pre-(62d0f012361a05e9a6a5bf71f58fe4d5b1b1df59) logic:


The original code still looks ok though, but as this was a change that was deliberately made by a Forge dev, I don't want to unconditionally revert it here.

**Comment by mezz** [comment]
Yours looks fine and does not reintroduce the bug, don't be afraid to replace existing code.

**Comment by bs2609** [comment]
So I checked out the context for that commit, this is how things look in build 1572, just before that commit was made:
forge_1572

It looks worse than it does now without those changes. Current vanilla has similar problems, see similar in 1.12.2:
vanilla

**Review by mezz — APPROVED** [review]

**Comment by mezz** [comment]
Tested and looks good to me. 
@bs2609 are you done tweaking this?

**Comment by bs2609** [comment]
Yeah, I think this is 'done' now.

**Comment by bs2609** [comment]
Another modpack test:

Numbers are for 10,807/21,614 `BakedItemModel` instances (equivalent, just depends on whether the culled 'extra' model is counted or not)

`UnpackedBakedQuad` numbers:

* Forge 2517: 509,758 instances, 640.9MB heap use
* Forge 2527: 388,485 instances, 488.5MB heap use

Total heap usage of `BakedItemModel` falls from 539.2MB to 401MB, saving 138.2MB (25.6%).

**Pull request merged**